src/ring_theory/valuation/valuation_subring.lean

```diff
@@ -361,4 +361,73 @@ variables {K} (A : valuation_subring K)
 lemma valuation_subring_valuation : A.valuation.valuation_subring = A :=
 by { ext, rw ← A.valuation_le_one_iff, refl }
 
+section unit_group
+
+/-- The unit group of a valuation subring, as a subgroup of `Kˣ`. -/
+def unit_group : subgroup Kˣ :=
+(A.valuation.to_monoid_with_zero_hom.to_monoid_hom.comp (units.coe_hom K)).ker
+
+lemma mem_unit_group_iff (x : Kˣ) : x ∈ A.unit_group ↔ A.valuation x = 1 := iff.refl _
+
+lemma unit_group_injective : function.injective (unit_group : valuation_subring K → subgroup _) :=
+λ A B, begin
+  rw [← A.valuation_subring_valuation, ← B.valuation_subring_valuation,
+    ← valuation.is_equiv_iff_valuation_subring,
+    A.valuation_subring_valuation, B.valuation_subring_valuation,
+    valuation.is_equiv_iff_val_eq_one, set_like.ext_iff],
+  intros h x,
```

Comment:
If you're rewriting the same thing in multiple directions, usually it means you weren't precise enough with your rewrite:
```suggestion
λ A B h, begin
  rw [← A.valuation_subring_valuation, ← B.valuation_subring_valuation,
    ← valuation.is_equiv_iff_valuation_subring, valuation.is_equiv_iff_val_eq_one],
  rw set_like.ext_iff at h,
  intros x,
```

Reply:
This is helpful advice, thank you